plugins/teststeps/teststeps.go

```diff
@@ -27,54 +29,94 @@ type PerTargetFunc func(cancel, pause <-chan struct{}, target *target.Target) er
-	for {
-		select {
-		case target := <-ch.In:
-			if target == nil {
-				// no more targets incoming
-				return nil
-			}
-			log.Debugf("%s: ForEachTarget: received target %s", pluginName, target)
-			errCh := make(chan error)
-			go func() {
-				log.Debugf("%s: ForEachTarget: calling function on target %s", pluginName, target)
-				errCh <- f(cancel, pause, target)
-			}()
+	type tgtErr struct {
+		target *target.Target
+		err    error
+	}
+	tgtErrCh := make(chan tgtErr)
+	tgtInFlight := int32(0)
+	noMoreTargetsCh := make(chan struct{})
 
+	go func(tgtInFlight *int32) {
+		defer func() {
+			log.Debugf("%s: ForEachTarget: exiting incoming loop, targets inflight %d", pluginName, tgtInFlight)
+		}()
+		for {
 			select {
-			case err := <-errCh:
-				if err != nil {
+			case tgt := <-ch.In:
+				log.Debugf("%s: ForEachTarget: received target %s", pluginName, tgt)
+				if tgt == nil {
+					log.Debugf("%s: ForEachTarget: all targets have been received", pluginName)
+					noMoreTargetsCh <- struct{}{}
+					return
+				}
+
+				go func() {
+					tgtErrCh <- tgtErr{target: tgt, err: f(cancel, pause, tgt)}
+				}()
+				atomic.AddInt32(tgtInFlight, 1)
+			case <-cancel:
+				log.Debugf("%s: ForEachTarget: incoming loop canceled", pluginName)
+				return
+			case <-pause:
+				log.Debugf("%s: ForEachTarget: incoming loop paused", pluginName)
+				return
+			}
+
+		}
+	}(&tgtInFlight)
+
+	noMoreTargets := false
+	reportResults := true
+	defer func() {
+		log.Debugf("%s: ForEachTarget: exiting outgoing loop, targets inflight %d, the last target received %v", pluginName, tgtInFlight, noMoreTargets)
+	}()
+	for {
+		select {
+		case te := <-tgtErrCh:
+			atomic.AddInt32(&tgtInFlight, -1)
+			if reportResults {
+				if te.err != nil {
 					select {
-					case ch.Err <- cerrors.TargetError{Target: target, Err: err}:
-						log.Errorf("%s: ForEachTarget: failed to apply test step function on target %s: %v", pluginName, target, err)
+					case ch.Err <- cerrors.TargetError{Target: te.target, Err: te.err}:
+						log.Errorf("%s: ForEachTarget: failed to apply test step function on target %s: %v", pluginName, te.target, te.err)
 					case <-cancel:
-						log.Debugf("%s: ForEachTarget: received cancellation signal", pluginName)
-						return nil
+						log.Debugf("%s: ForEachTarget: received cancellation signal while reporting error", pluginName)
+						reportResults = false
 					case <-pause:
-						log.Debugf("%s: ForEachTarget: received pausing signal", pluginName)
-						return nil
+						log.Debugf("%s: ForEachTarget: received pausing signal while reporting error", pluginName)
+						reportResults = false
 					}
 				} else {
 					select {
-					case ch.Out <- target:
-						log.Debugf("%s: ForEachTarget: target %s completed successfully", pluginName, target)
+					case ch.Out <- te.target:
+						log.Debugf("%s: ForEachTarget: target %s completed successfully", pluginName, te.target)
```

Comment:
**warning:**

This `Debugf` should be before `select`, if I understand correctly.